python/raydp/tf/estimator.py

```diff
@@ -75,168 +77,200 @@ def __init__(self,
+        self._metrics = metrics
+        self._feature_columns = feature_columns
+        self._feature_types = feature_types
+        self._feature_shapes = feature_shapes
+        self._label_column = label_column
+        self._label_shape = label_shape
+        self._label_type = label_type
+        self._batch_size = batch_size
+        self._num_epochs = num_epochs
+        self._shuffle = shuffle
+        self._callbacks = callbacks
+        self._extra_config = extra_config
+        self._trainer: Trainer = None
+        self._check()
 
-        # model
-        assert model is not None, "model must be not be None"
-        if isinstance(model, keras.Model):
-            self._serialized_model = model.to_json()
-        else:
-            raise Exception("Unsupported parameter, we only support tensorflow.keras.Model")
+    def _check(self):
+        assert self._model is not None, "Model must be provided"
+        assert isinstance(self._model, keras.Model), "Unsupported parameter, "\
+                                    "we only support tensorflow.keras.Model"
+        assert self._optimizer is not None, "Optimizer must be provided"
+        assert self._loss is not None, "Loss must be provided"
 
-        # optimizer
+    @staticmethod
+    def build_and_compile_model(config):
+        # create model
+        model: keras.Model = keras.models.model_from_json(config["model"])
+
+        # create optimizer
         # TODO: we should support multiple optimizers for multiple outputs model
-        assert optimizer is not None, "optimizer must not be None"
-        if isinstance(optimizer, str):
+        if isinstance(config["optimizer"], str):
             # it is a str represents the optimizer
-            _optimizer = optimizer
-        elif isinstance(optimizer, keras.optimizers.Optimizer):
-            _optimizer = keras.optimizers.serialize(optimizer)
+            serialized_optimizer = config["optimizer"]
+        elif isinstance(config["optimizer"], keras.optimizers.Optimizer):
+            serialized_optimizer = keras.optimizers.serialize(config["optimizer"])
         else:
             raise Exception(
                 "Unsupported parameter, we only support keras.optimizers.Optimizer subclass "
                 "instance or a str to represent the optimizer")
-        self._serialized_optimizer = _optimizer
+        optimizer = keras.optimizers.get(serialized_optimizer)
 
-        # loss
+        # create loss
         # TODO: we should support multiple losses for multiple outputs model
-        assert loss is not None, "loss must not be None"
-        if isinstance(loss, str):
-            _loss = loss
-        elif isinstance(loss, keras.losses.Loss):
-            _loss = keras.losses.serialize(loss)
+        if isinstance(config["loss"], str):
+            serialized_loss = config["loss"]
+        elif isinstance(config["loss"], keras.losses.Loss):
+            serialized_loss = keras.losses.serialize(config["loss"])
```

Comment:
Why do we need to serialize and deserialize the loss in the same method?

Reply:
Sry, I didn't realize this problem before, and I corrected it as previous method. Thx